This library contains a set of nonlinearities with first-order and second-order antialiasing functions

dariosanfilippo opened the pull request:
@sletz @josmithiii @orlarey @rmichon 

The library contains bounded and unbounded saturators as well as standard trigonometric functions with first-order and second-order aliasing suppression based on antiderivatives. Reference: https://www.dafx.de/paper-archive/2016/dafxpapers/20-DAFx-16_paper_41-PN.pdf.

Some of the second-order functions are not very accurate for a wide amplitude range in the input, but the first-order ones are solid. The authors provided examples of hard-clipping and tanh() functions, the most common saturators, but it is amazing that the technique can be applied to any waveshaping nonlinearities with an antiderivative.

What's also exciting is that linear combination of the antialiased functions can be effective to achieve antialiasing in nonlinearities that do not have an antiderivative, for example, sin(cos(x)).

The second-order antialiased functions for tanh() and tan were problematic, mostly for the fact that their antiderivative requires the dilogarithm function.

In a couple of cases, the antialiasing technique did not seem to be effective on a nonlinearity. We can decide whether to keep those for completeness or just have the working ones.

You can see the test here: https://www.dropbox.com/sh/1cmxwwcwr68vieu/AADjcLqEPFer82sEWRSVUZOBa?dl=0.

Tests have been performed at 192 kHz SR, in double-precision (necessary for the division by differences – we may need to add that in the comments). Except where otherwise noted, the input signal is a sinusoid at 4863 Hz and amplification of 256 units, about 48 dB gain.

I find this technique quite remarkable, especially in Faust since the second-order ADAA can be very effective with factors 2, 3, 4 oversampling, hence working at 96 kHz or 192 kHz SR.

Let me know if anything else is needed and if you also think that it can be a nice addition to the libraries.

Ciao,
Dario

dariosanfilippo commented:
Line 112: 

D1q = ba.if(abs(x - x') <= EPS, ill_D1q, safe_D1q) 

should be 

D1q = ba.if((x - x') ^ 2.0  <= EPS, ill_D1q, safe_D1q).

I'll fix that; hopefully, it will work even better.

josmithiii commented:
Here is where I found the paper: http://dafx16.vutbr.cz/dafxpapers/20-DAFx-16_paper_41-PN.pdf

dariosanfilippo commented:
Thanks, @josmithiii. Do you mean that the link that I provided was not working for you?

josmithiii commented:
Yes, the link failed for me, but it does work now.  Strange . . .

On Sun, Feb 14, 2021 at 5:47 AM Dario Sanfilippo <notifications@github.com>
wrote:

> Thanks, @josmithiii <https://github.com/josmithiii>. Do you mean that the
> link that I provided was not working for you?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/grame-cncm/faustlibraries/pull/70#issuecomment-778780377>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAQZKFMYW7XKZVJ2E6BVI73S67H5NANCNFSM4XSGZOAA>
> .
>


-- 
Julius O. Smith III <jos@ccrma.stanford.edu>
Professor of Music and, by courtesy, Electrical Engineering
CCRMA, Stanford University
http://ccrma.stanford.edu/~jos/

josmithiii commented:
I vote in favor of a new Faust Library on alias-suppressed nonlinearities

sletz commented:
BTW: adding a library means also completing the https://github.com/grame-cncm/faustlibraries/blob/master/all.lib and https://github.com/grame-cncm/faustlibraries/blob/master/stdfaust.lib files. @dariosanfilippo can you then compute the PR, thanks ?

dariosanfilippo commented:
@sletz Done. I've also added more information and I removed the functions that were not convincing, at least for now.

@josmithiii Correct me if I'm wrong, but it could be that cosh(x) and sinh(x) don't work as expected because they are each other's antiderivatives and their slope skyrockets pretty soon. However, I tested them with low-amplitude inputs (around 6 dB) and they well still not working, but that might just be that these ADAA algorithms are tricky in that case for the division by differences.
